File: easy-projects/easy-common/src/main/java/xyz/srclab/common/base/Hash.java
package xyz.srclab.common.base;

import xyz.srclab.annotation.Nullable;

import java.util.Arrays;
import java.util.Objects;

/**
 * @author sunqian
 */
public class Hash {

    public static int hash(@Nullable Object any) {
        if (any == null) {
            return 0;
        }
        if (any instanceof Object[]) {
            return Arrays.deepHashCode((Object[]) any);
        }
        if (any instanceof boolean[]) {
            return Arrays.hashCode((boolean[]) any);
        }
        if (any instanceof byte[]) {
            return Arrays.hashCode((byte[]) any);
        }
        if (any instanceof short[]) {
            return Arrays.hashCode((short[]) any);
        }
        if (any instanceof char[]) {
            return Arrays.hashCode((char[]) any);
        }
        if (any instanceof int[]) {
            return Arrays.hashCode((int[]) any);
        }
        if (any instanceof long[]) {
            return Arrays.hashCode((long[]) any);
        }
        if (any instanceof float[]) {
            return Arrays.hashCode((float[]) any);
        }
        if (any instanceof double[]) {
            return Arrays.hashCode((double[]) any);
        }
        return Objects.hashCode(any);
    }

    public static int hash(Object... objects) {
        return Arrays.deepHashCode(objects);
    }
}

File: srclab-common/src/main/kotlin/xyz/srclab/common/base/Hash.kt
@file:JvmName("Hash")
@file:JvmMultifileClass

package xyz.srclab.common.base

import java.util.*

fun Any?.hash(): Int {
    return this.hashCode()
}

fun Any?.elementHash(): Int {
    if (this === null) {
        return this.hashCode()
    }
    return when (this) {
        is BooleanArray -> Arrays.hashCode(this)
        is ShortArray -> Arrays.hashCode(this)
        is CharArray -> Arrays.hashCode(this)
        is IntArray -> Arrays.hashCode(this)
        is LongArray -> Arrays.hashCode(this)
        is FloatArray -> Arrays.hashCode(this)
        is DoubleArray -> Arrays.hashCode(this)
        is Array<*> -> Arrays.hashCode(this)
        else -> this.hashCode()
    }
}

fun Any?.elementDeepHash(): Int {
    if (this === null) {
        return this.hashCode()
    }
    return when (this) {
        is BooleanArray -> Arrays.hashCode(this)
        is ShortArray -> Arrays.hashCode(this)
        is CharArray -> Arrays.hashCode(this)
        is IntArray -> Arrays.hashCode(this)
        is LongArray -> Arrays.hashCode(this)
        is FloatArray -> Arrays.hashCode(this)
        is DoubleArray -> Arrays.hashCode(this)
        is Array<*> -> Arrays.deepHashCode(this)
        else -> this.hashCode()
    }
}

fun hash(vararg args: Any?): Int {
    return args.contentHashCode()
}

fun deepHash(vararg args: Any?): Int {
    return args.contentDeepHashCode()
}